File: mini-markov-python/mini_markov/chain.py
"""Generic Markov Chain implementation."""

from typing import TypeVar, Generic, Dict, List, Optional, Tuple
from collections import defaultdict
import random
import math

T = TypeVar("T")


class MarkovChain(Generic[T]):
    """
    A generic Markov chain implementation supporting any hashable type.
    
    Attributes:
        order: The number of previous states to consider (n-gram size).
    """
    
    def __init__(self, order: int = 1):
        """
        Creates a new Markov chain.
        
        Args:
            order: Number of previous states to consider (must be >= 1).
        """
        if order < 1:
            raise ValueError("Order must be at least 1")
        
        self.order = order
        self._transitions: Dict[Tuple[T, ...], Dict[T, int]] = defaultdict(lambda: defaultdict(int))
        self._start_states: Dict[Tuple[T, ...], int] = defaultdict(int)
        self._num_sequences = 0
    
    def train(self, sequence: List[T]) -> None:
        """
        Trains the chain on a sequence of states.
        
        Args:
            sequence: The sequence to train on.
        """
        if len(sequence) <= self.order:
            return
        
        self._num_sequences += 1
        
        # Record start state
        start = tuple(sequence[:self.order])
        self._start_states[start] += 1
        
        # Record transitions
        for i in range(len(sequence) - self.order):
            current = tuple(sequence[i:i + self.order])
            next_state = sequence[i + self.order]
            self._transitions[current][next_state] += 1
    
    def train_many(self, sequences: List[List[T]]) -> None:
        """Trains the chain on multiple sequences."""
        for seq in sequences:
            self.train(seq)
    
    def get_probabilities(self, current: List[T]) -> Optional[Dict[T, float]]:
        """
        Gets the probability distribution for next states.
        
        Args:
            current: The current state sequence (must have length == order).
            
        Returns:
            Dictionary mapping states to probabilities, or None if unknown state.
        """
        if len(current) != self.order:
            return None
        
        key = tuple(current)
        if key not in self._transitions:
            return None
        
        counts = self._transitions[key]
        total = sum(counts.values())
        
        return {state: count / total for state, count in counts.items()}
    
    def sample_next(self, current: List[T], rng: random.Random = None) -> Optional[T]:
        """
        Samples the next state given the current state.
        
        Args:
            current: The current state sequence.
            rng: Random number generator (uses default if None).
            
        Returns:
            The sampled next state, or None if unknown state.
        """
        if rng is None:
            rng = random.Random()
            
        if len(current) != self.order:
            return None
        
        key = tuple(current)
        if key not in self._transitions:
            return None
        
        counts = self._transitions[key]
        total = sum(counts.values())
        
        threshold = rng.randint(0, total - 1)
        
        for state, count in counts.items():
            threshold -= count
            if threshold < 0:
                return state
        
        return next(iter(counts.keys()))
    
    def sample_start(self, rng: random.Random = None) -> Optional[List[T]]:
        """Samples a random starting state."""
        if rng is None:
            rng = random.Random()
            
        if not self._start_states:
            return None
        
        total = sum(self._start_states.values())
        threshold = rng.randint(0, total - 1)
        
        for start, count in self._start_states.items():
            threshold -= count
            if threshold < 0:
                return list(start)
        
        return list(next(iter(self._start_states.keys())))
    
    def generate(self, length: int, rng: random.Random = None) -> List[T]:
        """
        Generates a sequence of states.
        
        Args:
            length: Maximum length of the generated sequence.
            rng: Random number generator.
            
        Returns:
            Generated sequence of states.
        """
        if rng is None:
            rng = random.Random()
            
        start = self.sample_start(rng)
        if start is None:
            return []
        
        result = list(start)
        
        while len(result) < length:
            current = result[-self.order:]
            next_state = self.sample_next(current, rng)
            if next_state is None:
                break
            result.append(next_state)
        
        return result
    
    @property
    def num_states(self) -> int:
        """Returns the number of unique state sequences."""
        return len(self._transitions)
    
    @property
    def num_transitions(self) -> int:
        """Returns the total number of transitions recorded."""
        return sum(sum(counts.values()) for counts in self._transitions.values())
    
    @property
    def num_sequences(self) -> int:
        """Returns the number of training sequences."""
        return self._num_sequences
    
    def entropy(self) -> float:
        """Calculates the entropy of the chain in bits."""
        total_entropy = 0.0
        total_weight = 0.0
        
        for counts in self._transitions.values():
            total = sum(counts.values())
            if total == 0:
                continue
            
            state_entropy = 0.0
            for count in counts.values():
                if count > 0:
                    p = count / total
                    state_entropy -= p * math.log2(p)
            
            total_entropy += state_entropy * total
            total_weight += total
        
        return total_entropy / total_weight if total_weight > 0 else 0.0
    
    def clear(self) -> None:
        """Clears all learned transitions."""
        self._transitions.clear()
        self._start_states.clear()
        self._num_sequences = 0
    
    def __repr__(self) -> str:
        return f"MarkovChain(order={self.order}, states={self.num_states}, transitions={self.num_transitions})"

File: mini-markov-go/markov/chain.go
// Package markov provides a generic Markov chain implementation.
package markov

import (
	"fmt"
	"math"
	"math/rand"
	"strings"
)

// Chain represents a generic Markov chain.
type Chain struct {
	Order        int
	Transitions  map[string]map[string]int
	StartStates  map[string]int
	NumSequences int
}

// NewChain creates a new Markov chain with the specified order.
func NewChain(order int) *Chain {
	if order < 1 {
		panic("Order must be at least 1")
	}
	return &Chain{
		Order:       order,
		Transitions: make(map[string]map[string]int),
		StartStates: make(map[string]int),
	}
}

// stateKey converts a slice of states to a string key.
func stateKey(states []string) string {
	return strings.Join(states, "\x00")
}

// Train trains the chain on a sequence of states.
func (c *Chain) Train(sequence []string) {
	if len(sequence) <= c.Order {
		return
	}

	c.NumSequences++

	// Record start state
	start := stateKey(sequence[:c.Order])
	c.StartStates[start]++

	// Record transitions
	for i := 0; i <= len(sequence)-c.Order-1; i++ {
		current := stateKey(sequence[i : i+c.Order])
		next := sequence[i+c.Order]

		if c.Transitions[current] == nil {
			c.Transitions[current] = make(map[string]int)
		}
		c.Transitions[current][next]++
	}
}

// TrainMany trains on multiple sequences.
func (c *Chain) TrainMany(sequences [][]string) {
	for _, seq := range sequences {
		c.Train(seq)
	}
}

// GetProbabilities returns the probability distribution for next states.
func (c *Chain) GetProbabilities(current []string) map[string]float64 {
	if len(current) != c.Order {
		return nil
	}

	key := stateKey(current)
	counts, exists := c.Transitions[key]
	if !exists {
		return nil
	}

	total := 0
	for _, count := range counts {
		total += count
	}

	probs := make(map[string]float64)
	for state, count := range counts {
		probs[state] = float64(count) / float64(total)
	}
	return probs
}

// SampleNext samples the next state given the current state.
func (c *Chain) SampleNext(current []string, rng *rand.Rand) (string, bool) {
	if len(current) != c.Order {
		return "", false
	}

	key := stateKey(current)
	counts, exists := c.Transitions[key]
	if !exists {
		return "", false
	}

	total := 0
	for _, count := range counts {
		total += count
	}

	threshold := rng.Intn(total)
	for state, count := range counts {
		threshold -= count
		if threshold < 0 {
			return state, true
		}
	}

	// Should never reach here
	for state := range counts {
		return state, true
	}
	return "", false
}

// SampleStart samples a random starting state.
func (c *Chain) SampleStart(rng *rand.Rand) ([]string, bool) {
	if len(c.StartStates) == 0 {
		return nil, false
	}

	total := 0
	for _, count := range c.StartStates {
		total += count
	}

	threshold := rng.Intn(total)
	for key, count := range c.StartStates {
		threshold -= count
		if threshold < 0 {
			return strings.Split(key, "\x00"), true
		}
	}

	for key := range c.StartStates {
		return strings.Split(key, "\x00"), true
	}
	return nil, false
}

// Generate generates a sequence of states.
func (c *Chain) Generate(length int, rng *rand.Rand) []string {
	start, ok := c.SampleStart(rng)
	if !ok {
		return nil
	}

	result := make([]string, len(start))
	copy(result, start)

	for len(result) < length {
		current := result[len(result)-c.Order:]
		next, ok := c.SampleNext(current, rng)
		if !ok {
			break
		}
		result = append(result, next)
	}

	return result
}

// NumStates returns the number of unique state sequences.
func (c *Chain) NumStates() int {
	return len(c.Transitions)
}

// NumTransitions returns the total number of transitions recorded.
func (c *Chain) NumTransitions() int {
	total := 0
	for _, counts := range c.Transitions {
		for _, count := range counts {
			total += count
		}
	}
	return total
}

// Entropy calculates the entropy of the chain.
func (c *Chain) Entropy() float64 {
	totalEntropy := 0.0
	totalWeight := 0.0

	for _, counts := range c.Transitions {
		total := 0
		for _, count := range counts {
			total += count
		}
		if total == 0 {
			continue
		}

		stateEntropy := 0.0
		for _, count := range counts {
			if count > 0 {
				p := float64(count) / float64(total)
				stateEntropy -= p * math.Log2(p)
			}
		}

		totalEntropy += stateEntropy * float64(total)
		totalWeight += float64(total)
	}

	if totalWeight > 0 {
		return totalEntropy / totalWeight
	}
	return 0.0
}

// String returns a string representation of the chain.
func (c *Chain) String() string {
	return fmt.Sprintf("MarkovChain(order=%d, states=%d, transitions=%d)",
		c.Order, c.NumStates(), c.NumTransitions())
}
